### ISSUE TITLE
VRAM compressed textures import at extremely low quality

### ISSUE THREAD

**Issue by Interference** [godotengine/godot#100114]
### Tested versions

v4.3.stable.official [77dcf97d8]

### System information

Windows 10 - Godot v4.3 stable - Vulkan Forward+ renderer. nVidia GTX 1080. 16GB RAM. Intel Core i5-4690K CPU @ 3.50GHz.

### Issue description

Importing textures intended for 3D models using the default VRAM compressed mode (with "high quality" turned off) produces extremely low quality results; worse than they should be. This is particularly noticeable with textures 1024 pixels square and lower.

As an example, below is a container prop with an ORM material applied and 512 x 512 textures. Godot has converted the ORM and base colour to DXT1 and the normal map to RGTC. The texture should appear as mostly yellowy orange with some light edge wear exposing the underlying metal. Due to the low quality of the texture compression, however, the edge wear displays as large, blocky green blotches along all the mesh edges.

![Image](https://github.com/user-attachments/assets/9a6f595d-9457-4a1b-ae50-2a395ecbaf89)

Below is the original source file for the base colour, in PNG format.

![Image](https://github.com/user-attachments/assets/fb1365cd-a32c-484e-9edc-a87b61e40cd6)

For comparison, here's the same texture run through nVidia's DXT Tools at the default settings. The below image was saved to DDS, then converted to PNG for upload.

![Image](https://github.com/user-attachments/assets/b4909cad-aab3-4c6a-a6eb-0b2785d2325d)

Note how the compression artifacts are barely visible.

Below is another example, this time run through an online image converter called [Convertio](https://convertio.co/dds-converter/). While the quality is not quite as good as nVidia's, the results are significantly better than Godot's own.

![Image](https://github.com/user-attachments/assets/852ab0b0-20e8-474d-af32-a78d85b2e45c)

Here is the container again, only this time with the base colour texture replaced with the Convertio version. Notice how the green discolouration is completely gone.

![Image](https://github.com/user-attachments/assets/56bbf629-51d7-4773-98fb-765c0be94bed)

### Steps to reproduce

1. Right-click and download the above source base colour texture for the crate. Drop it into a Godot project and set the compress mode to VRAM compressed
2. Create a MeshInstance3D, assign it a plane mesh, a material, and assign this texture as the base colour
3. Use any suitable image converter (Convertio is a good choice if you wish to exactly replicate my results) to convert the PNG to DDS yourself, then add the DDS to the project
4. Create a new MeshInstance3D, assign it another plane mesh and material then assign this new texture to the base colour
5. Observe the quality difference between the two examples

### Minimal reproduction project (MRP)

N/A

**Comment by fire** [comment]
Are you able to turn on BPCT or ASTC quality mode?

**Comment by Interference** [comment]
> Are you able to turn on BPCT or ASTC quality mode? The dxt5 and etc2 modes are tuned for speed and not quality.

Yes, although those present another problem: the current implementation of BPCT in Godot, at least for files without an alpha channel, are nearly twice the size of their DXT1 counterparts. I believe this is ALSO a bug, since they list the textures as RGBA even though they don't have an alpha channel, meaning the conversion process is treating them as if they have one.

That said, I'm fully aware DXT compression is of a lower quality but it shouldn't be THIS low. That much is evident when doing the compression yourself rather than leaving it to Godot, something I'd rather not have to do since the whole point of having Godot handle it is to speed the process of getting art into the engine.

**Comment by fire** [comment]
I revisited the texture compression code for etc2 and dxt5; there's no tunable for etcpak.

https://github.com/godotengine/godot/blob/master/modules/etcpak/image_compress_etcpak.cpp#L256

However, Godot Engine has Betsy, which does fast BPC encoding without importing the standard ISPC library dependency.

In what direction do you want Godot Engine to optimize so I can estimate the options?

**Comment by fire** [comment]
1. I don't know if we can rewrite or upgrade etcpak to have higher quality
2. Importing [ISPCTextureCompressor](https://github.com/GameTechDev/ISPCTextureCompressor) is not wanted because of the ispc dependency. When I investigate 5 years ago, it was an external process.
3. We could set Betsy as the default for better texture compression quality.

**Comment by Interference** [comment]
> I revisited the texture compression code for etc2 and dxt5; there's no tunable for etcpak.
> 
> https://github.com/godotengine/godot/blob/master/modules/etcpak/image_compress_etcpak.cpp#L256
> 
> However, Godot Engine has Betsy, which does fast BPTC encoding without importing the standard ISPC library dependency.
> 
> In what direction do you want Godot Engine to optimize so I can estimate the options?

The ideal solution would be if you were able to tweak the DXT compression functionality to increase its quality to something broadly comparable to the image uploader example I linked. Even just reducing the blocky discolouration would help. From your response, it sounds like that would be a challenge, though.

While it's not strictly a fix in of itself, I wouldn't have an issue with switching to BPCT compression to side step the issue if the files weren't so large.

Does conversion to BC7 only output files with an alpha channel? If I am correct and there is a variant that does NOT include an alpha channel, are textures encoded in it notably smaller? If they are, would you be able to make Godot's texture importer detect image files without an alpha channel and use that variation instead?

**Comment by fire** [comment]
@BlueCube3310 thought?

**Comment by Interference** [comment]
I put my research hat on and had a look into BC7 and, from what I can tell at least, it seems like you get an alpha channel with it whether you like it or not, meaning I don't think there's an avenue for reducing BC7 texture file sizes.

My current workaround is as follows:

1. For textures with transparency and normal maps I set "High Quality" and import them as BC7
2. For textures with no transparency I manually convert them to DXT1 using Nvidia's Texture Tools Exporter and drop the results into my project

This isn't exactly ideal as there's an extra step each time a texture is altered (plus you lose a few other useful features) but the quality difference is night and day and and the files are half the size of BC7.

**Comment by fire** [comment]
bc6 might be the one without alpha

**Comment by clayjohn** [comment]
> bc6 might be the one without alpha

BC6H is RGB HDR. BC7 is RGBA LDR. And yes, BC7 will always be twice the size of DXT1. 

We don't have a GPU compressor for BC7 right now, so it is also very slow to import.

**Comment by Interference** [comment]
> bc6 might be the one without alpha

I experimented with that in Texture Tools. It lists two variations, signed and unsigned, both without alpha but it seems to produce only slightly smaller files than BC7 (341kb rather than around 350kb for a 512x512 image).

**Comment by clayjohn** [comment]
The difference is just header data.  They are fixed format compression schemes, so each 4x4 block always has the same size

I just tested that PNG locally and I am not getting the green color. I am using master, so I will try again in 4.3 to check if this is something that has already been fixed

**Comment by clayjohn** [comment]
Just confirmed, this is fixed in 4.4:

_4.3_
![Image](https://github.com/user-attachments/assets/ff40b4ad-6a77-4d02-aa9e-bc523e1250ba)

_4.4 lossless_
![Image](https://github.com/user-attachments/assets/2b8fffb7-7dc7-4560-b9fd-74ea4c317678)

_4.4 DXT1_
![Image](https://github.com/user-attachments/assets/1949fadd-ec76-4999-a425-049e1d0f6fe8)

**Comment by BlueCube3310** [comment]
The discolouring is likely caused by etcpak dithering the images in an attempt to increase the quality. Without the dithering, blocky patterns become very apparent so it's a double-edged sword. With 4.4 and Betsy the quality should be better by default

**Comment by clayjohn** [comment]
> The discolouring is likely caused by etcpak dithering the images in an attempt to increase the quality. Without the dithering, blocky patterns become very apparent so it's a double-edged sword. With 4.4 and Betsy the quality should be better by default

I just checked. Indeed using etcpak in 4.4 brings the problem back. It is fixed by using Betsy

**Comment by Interference** [comment]
@clayjohn That's great news. Roll on 4.4!
